src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,303 @@
+                await this.textWriter.DisposeAsync();
+                await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Whether the current TextWriter is writing JSON
+        /// </summary>
+        /// <returns></returns>
+        private bool IsWritingJson
+        {
+            get
+            {
+                return String.IsNullOrEmpty(this.currentTextWriterContentType) || this.currentTextWriterContentType.StartsWith(MimeConstants.MimeApplicationJson, StringComparison.Ordinal);
+            }
+        }
+
+        /// <summary>
+        /// Represents a TextWriter implementation for writing UTF-8 JSON data in an OData context.
+        /// </summary>
+        internal class ODataUtf8JsonTextWriter : TextWriter
+        {
+            ODataUtf8JsonWriter jsonWriter = null;
+            private char[] buffer;
+            int numOfCharsNotWrittenFromPreviousChunk = 0;
+
+            public ODataUtf8JsonTextWriter(ODataUtf8JsonWriter jsonWriter)
+            {
+                this.jsonWriter = jsonWriter;
+            }
+
+            public override Encoding Encoding => throw new NotImplementedException();
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream synchronously.
+            /// </summary>
+            public override void Flush()
+            {
+                // If there are unprocessed chars, encode and write them as the final block.
+                ReadOnlySpan<char> charsNotProcessedFromPreviousChunk = this.buffer.AsSpan().Slice(numOfCharsNotWrittenFromPreviousChunk);
+
+                if (!charsNotProcessedFromPreviousChunk.IsEmpty)
+                {
+                    this.jsonWriter.WriteStringChunk(charsNotProcessedFromPreviousChunk, true);
+
+                    // set numOfCharsNotWrittenFromPreviousChunk to 0
+                    this.numOfCharsNotWrittenFromPreviousChunk = 0;
+                }
+
+                if (this.buffer != null)
+                {
+                    ArrayPool<char>.Shared.Return(this.buffer);
+                }
+
+                this.jsonWriter.Flush();
+            }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream asynchronously.
+            /// </summary>
+            /// <param name="cancellationToken">A cancellation token to observe while waiting for the flush operation to complete.</param>
+            /// <returns>A task representing the asynchronous flush operation.</returns>
+            public override async Task FlushAsync()
+            {
+                // If there are unprocessed chars, encode and write them as the final block.
+                ReadOnlyMemory<char> charsNotProcessedFromPreviousChunk = this.buffer.AsMemory().Slice(0, numOfCharsNotWrittenFromPreviousChunk);
+
+                if (!charsNotProcessedFromPreviousChunk.IsEmpty)

```

Comment:
See: https://github.com/OData/odata.net/pull/2880/files#r1535044028

Reply:
moved to dispose